diff --git a/core/modules/media/src/OEmbed/ResourceFetcher.php b/core/modules/media/src/OEmbed/ResourceFetcher.php
index 76c2d8b3b7f11da3b77632dc54c842a13410f67a..be325adfabd680c06f0f5dd8417ced8d77212fd9 100644
--- a/core/modules/media/src/OEmbed/ResourceFetcher.php
+++ b/core/modules/media/src/OEmbed/ResourceFetcher.php
@@ -9,7 +9,6 @@
 use Psr\Http\Client\ClientExceptionInterface;
 
 // cspell:ignore nocdata
-
 /**
  * Fetches and caches oEmbed resources.
  */
@@ -90,11 +89,129 @@ public function fetchResource($url) {
       throw new ResourceException('The oEmbed resource could not be decoded.', $url);
     }
 
+    $this->convertToHighResolutionThumbnail($data);
+
     $this->cacheBackend->set($cache_id, $data);
 
     return $this->createResource($data, $url);
   }
 
+  /**
+   * Convert the embed data to a high resolution thumbnail.
+   *
+   * @param array $data
+   *   The embed data altered by reference.
+   *
+   * @todo Convert to a hook or event for others to update to a high resolution.
+   * @todo Set the width and height of the high resolution image.
+   */
+  protected function convertToHighResolutionThumbnail(array &$data): void {
+    $provider_name = $data['provider_name'] ?? NULL;
+    $high_resolution_thumbnail_url = NULL;
+    if ($provider_name === 'Vimeo') {
+      $high_resolution_thumbnail_url = $this->getVimeoHighResolutionThumbnailUrl($data);
+    }
+    elseif ($provider_name === 'YouTube') {
+      $high_resolution_thumbnail_url = $this->getYouTubeHighResolutionThumbnailUrl($data);
+    }
+
+    if ($high_resolution_thumbnail_url) {
+      $data['thumbnail_url'] = $high_resolution_thumbnail_url;
+      $size = getimagesize($data['thumbnail_url']);
+      if ($size && count($size) > 1) {
+        $data['thumbnail_width'] = $size[0];
+        $data['thumbnail_height'] = $size[1];
+      }
+    }
+  }
+
+  /**
+   * Get the YouTube high resolution thumbnail.
+   *
+   * @param array $data
+   *   The embed data.
+   *
+   * @return string|null
+   *   The URL for the high resolution thumbnail.
+   */
+  protected function getYouTubeHighResolutionThumbnailUrl(array $data): ?string {
+    if (empty($data['thumbnail_url'])) {
+      return NULL;
+    }
+
+    $high_quality_thumbnail = preg_replace('@(hq)(\w+\.[^\.]+)$@Ui', 'maxres$2', $data['thumbnail_url']);
+    if ($high_quality_thumbnail === $data['thumbnail_url']) {
+      return NULL;
+    }
+
+    try {
+      $response = $this->httpClient->request('GET', $high_quality_thumbnail, [
+        RequestOptions::TIMEOUT => 5,
+        RequestOptions::HTTP_ERRORS => FALSE,
+      ]);
+
+      if ($response->getStatusCode() === 200) {
+        return $high_quality_thumbnail;
+      }
+    }
+    catch (\Throwable $e) {
+      // Any error, use the default thumbnail.
+    }
+
+    return NULL;
+  }
+
+  /**
+   * Get the Vimeo high resolution thumbnail.
+   *
+   * @param array $data
+   *   The embed data.
+   *
+   * @return string|null
+   *   The URL for the high resolution thumbnail.
+   */
+  protected function getVimeoHighResolutionThumbnailUrl(array $data): ?string {
+    if (!empty($data['video_id'])) {
+      $vimeo_url = 'https://vimeo.com/api/v2/video/' . $data['video_id'] . '.xml';
+      try {
+        $response = $this->httpClient->request('GET', $vimeo_url, [
+          RequestOptions::TIMEOUT => 5,
+          RequestOptions::HTTP_ERRORS => FALSE,
+        ]);
+        $content = (string) $response->getBody();
+        $parsed_data = $this->parseResourceXml($content, $vimeo_url);
+        if ($parsed_data &&
+          !empty($parsed_data['video']['thumbnail_large']) &&
+          is_string($parsed_data['video']['thumbnail_large'])) {
+          return (string) $parsed_data['video']['thumbnail_large'];
+        }
+      }
+      catch (\Throwable $e) {
+        // Any error use the default thumbnail.
+      }
+    }
+
+    // The video might've been private, so not available via the API.  But can
+    // we extract a higher quality URL from the thumbnail URL.
+    if (!empty($data['thumbnail_url'])) {
+      $vimeo_url = preg_replace('/(.*)(_\d+(x\d+)?)/m', '$1', $data['thumbnail_url']);
+      try {
+        $response = $this->httpClient->request('GET', $vimeo_url, [
+          RequestOptions::TIMEOUT => 5,
+          RequestOptions::HTTP_ERRORS => FALSE,
+        ]);
+        $statusCode = $response->getStatusCode();
+        if ($statusCode >= 200 && $statusCode < 300) {
+          return $vimeo_url;
+        }
+      }
+      catch (\Throwable $e) {
+        // Any error use the default thumbnail.
+      }
+    }
+    return NULL;
+  }
+
   /**
    * Creates a Resource object from raw resource data.
    *
