diff --git a/core/modules/media_library/js/media_library.widget.js b/core/modules/media_library/js/media_library.widget.js
index b9f5b8a5cb17d8aaff7fa8ac863e8d3f0d15ec25..6205288c16787702fcf263f3add96b91b3e578a6 100644
--- a/core/modules/media_library/js/media_library.widget.js
+++ b/core/modules/media_library/js/media_library.widget.js
@@ -72,6 +72,28 @@
     },
   };
 
+  /**
+   * Allow users to edit media library items inside a modal.
+   */
+  Drupal.behaviors.MediaLibraryWidgetEditItem = {
+    attach: function attach(context) {
+      const mediaLibraryEdit = once(
+        'media-library-edit',
+        '.media-library-widget .edit-media',
+        context,
+      );
+      $(mediaLibraryEdit).on('click', function () {
+        // Remove any "selected-media" classes.
+        $(this)
+          .closest('.media-library-selection')
+          .find('.selected-media')
+          .removeClass('selected-media');
+        // Mark the media item as selected to render it properly when submitting an ajax media edit request.
+        $(this).parent().find('article').addClass('selected-media');
+      });
+    },
+  };
+
   /**
    * Disable the open button when the user is not allowed to add more items.
    *
diff --git a/core/modules/media_library/media_library.module b/core/modules/media_library/media_library.module
index 81c4f7900b6c25175245c1e0ad797df5722b32bf..aa8c8891ca6ca9da8d39fdee2ffa05e900ab9efd 100644
--- a/core/modules/media_library/media_library.module
+++ b/core/modules/media_library/media_library.module
@@ -4,13 +4,24 @@
  * @file
  */
 
+use Drupal\Component\Utility\Html;
+use Drupal\Core\Ajax\AjaxResponse;
+use Drupal\Core\Ajax\CloseModalDialogCommand;
+use Drupal\Core\Ajax\InvokeCommand;
+use Drupal\Core\Ajax\ReplaceCommand;
+use Drupal\Core\Entity\EntityFormInterface;
 use Drupal\Core\Entity\Entity\EntityFormDisplay;
 use Drupal\Core\Entity\Entity\EntityViewDisplay;
+use Drupal\Core\EventSubscriber\MainContentViewSubscriber;
 use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Messenger\MessengerInterface;
+use Drupal\Core\Render\Element;
 use Drupal\image\Entity\ImageStyle;
 use Drupal\image\Plugin\Field\FieldType\ImageItem;
 use Drupal\media\MediaTypeInterface;
 use Drupal\media_library\Hook\MediaLibraryThemeHooks;
+use Drupal\media_library\MediaLibraryState;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
 
 /**
  * Prepares variables for the media library modal dialog.
@@ -69,6 +80,106 @@ function _media_library_views_form_media_library_after_build(array $form, FormSt
   return $form;
 }
 
+/**
+ * Implements hook_form_FORM_ID_alter() for media_form forms.
+ */
+function media_library_form_media_form_alter(array &$form, FormStateInterface $form_state): void {
+  // Make the media edit form work inside modal dialogs.
+  $form_object = $form_state->getFormObject();
+  if (!$form_object instanceof EntityFormInterface || $form_object->getOperation() !== 'edit') {
+    return;
+  }
+
+  try {
+    $state = MediaLibraryState::fromRequest(\Drupal::request());
+  }
+  catch (BadRequestHttpException) {
+    return;
+  }
+  catch (\InvalidArgumentException) {
+    return;
+  }
+  if ($state->getOpenerId() !== 'media_library.opener.field_widget') {
+    return;
+  }
+
+  $wrapper_format = \Drupal::request()->query->get(MainContentViewSubscriber::WRAPPER_FORMAT);
+  if ($wrapper_format !== 'drupal_modal' && $wrapper_format !== 'drupal_ajax') {
+    return;
+  }
+
+  foreach (Element::children($form['actions']) as $key) {
+    if ($key === 'submit') {
+      // Add an AJAX submit handler to the 'Save' button.
+      $form['actions'][$key]['#ajax'] = [
+        'callback' => '_media_library_media_form_ajax_submit',
+      ];
+
+      // Prevent the form that opened the modal dialog to refocus to the
+      // 'Save' button at the bottom of the form.
+      $form['actions'][$key]['#attributes']['data-disable-refocus'] = 'true';
+
+      // The data-drupal-selector needs to be the same between the various
+      // AJAX requests. A bug in \Drupal\Core\Form\FormBuilder prevents that
+      // from happening unless $form['#id'] is also the same. Normally, #id
+      // is set to a unique HTML ID via Html::getUniqueId(), but here we
+      // bypass that in order to work around the data-drupal-selector bug.
+      // This is okay so long as we assume that this form only ever occurs
+      // once on a page.
+      // @todo Remove once https://www.drupal.org/node/2897377 is fixed.
+      $form['#id'] = Html::getId($form_state->getBuildInfo()['form_id']);
+    }
+    else {
+      // We just want to allow the save action for now. Just hide other
+      // actions like the delete action.
+      $form['actions'][$key]['#access'] = FALSE;
+    }
+  }
+}
+
+/**
+ * AJAX callback for media edit form when form is shown inside modal dialog.
+ *
+ * @ingroup form
+ */
+function _media_library_media_form_ajax_submit(array &$form, FormStateInterface $form_state): AjaxResponse {
+  $response = new AjaxResponse();
+
+  if ($form_state->hasAnyErrors()) {
+    $form['status_messages'] = [
+      '#type' => 'status_messages',
+      '#weight' => -1000,
+    ];
+    $response->addCommand(new ReplaceCommand('[data-drupal-selector="' . $form['#attributes']['data-drupal-selector'] . '"]', $form));
+
+    return $response;
+  }
+
+  try {
+    $state = MediaLibraryState::fromRequest(\Drupal::request());
+  }
+  catch (BadRequestHttpException) {
+    return $response;
+  }
+  catch (\InvalidArgumentException) {
+    return $response;
+  }
+
+  $opener_parameters = $state->getOpenerParameters();
+  if (empty($opener_parameters['field_widget_id'])) {
+    return $response;
+  }
+
+  // Trigger the field widget to update itself and close the dialog.
+  $response->addCommand(new InvokeCommand("[data-media-library-widget-update=\"{$opener_parameters['field_widget_id']}\"]", 'trigger', ['mousedown']));
+  $response->addCommand(new CloseModalDialogCommand());
+
+  // Remove status messages when editing medias.
+  \Drupal::messenger()->deleteByType(MessengerInterface::TYPE_STATUS);
+
+  return $response;
+}
+
 /**
  * Submit callback for media type form.
  */
diff --git a/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php b/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php
index 41c028be3a12dd636643529dc7ef8afe6f3e0c5c..107708400fe07af9e03a347ffcb250f7abfcd613 100644
--- a/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php
+++ b/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php
@@ -377,6 +377,55 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
       ];
     }
 
+    $cardinality_unlimited = ($element['#cardinality'] === FieldStorageDefinitionInterface::CARDINALITY_UNLIMITED);
+    $remaining = $element['#cardinality'] - count($referenced_entities);
+
+    // Inform the user of how many items are remaining.
+    if (!$cardinality_unlimited) {
+      if ($remaining) {
+        $cardinality_message = $this->formatPlural($remaining, 'One media item remaining.', '@count media items remaining.');
+      }
+      else {
+        $cardinality_message = $this->t('The maximum number of media items have been selected.');
+      }
+
+      // Add a line break between the field message and the cardinality message.
+      if (!empty($element['#description'])) {
+        $element['#description'] .= '<br />';
+      }
+      $element['#description'] .= $cardinality_message;
+    }
+
+    // Create a new media library URL with the correct state parameters.
+    $selected_type_id = reset($allowed_media_type_ids);
+    $remaining = $cardinality_unlimited ? FieldStorageDefinitionInterface::CARDINALITY_UNLIMITED : $remaining;
+    // This particular media library opener needs some extra metadata for its
+    // \Drupal\media_library\MediaLibraryOpenerInterface::getSelectionResponse()
+    // to be able to target the element whose 'data-media-library-widget-value'
+    // attribute is the same as $field_widget_id. The entity ID, entity type ID,
+    // bundle, field name are used for access checking.
+    $entity = $items->getEntity();
+    $opener_parameters = [
+      'field_widget_id' => $field_widget_id,
+      'entity_type_id' => $entity->getEntityTypeId(),
+      'bundle' => $entity->bundle(),
+      'field_name' => $field_name,
+    ];
+    if (isset($form['#form_mode'])) {
+      $opener_parameters['form_mode'] = $form['#form_mode'];
+    }
+    // Only add the entity ID when we actually have one. The entity ID needs to
+    // be a string to ensure that the media library state generates its
+    // tamper-proof hash in a consistent way.
+    if (!$entity->isNew()) {
+      $opener_parameters['entity_id'] = (string) $entity->id();
+
+      if ($entity->getEntityType()->isRevisionable()) {
+        $opener_parameters['revision_id'] = (string) $entity->getRevisionId();
+      }
+    }
+    $state = MediaLibraryState::create('media_library.opener.field_widget', $allowed_media_type_ids, $selected_type_id, $remaining, $opener_parameters);
+
     $element['selection'] = [
       '#type' => 'container',
       '#theme_wrappers' => [
@@ -425,6 +474,7 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
           '#type' => 'submit',
           '#name' => $field_name . '-' . $delta . '-media-library-remove-button' . $id_suffix,
           '#value' => $this->t('Remove'),
+          '#weight' => -10,
           '#media_id' => $media_item->id(),
           '#attributes' => [
             'aria-label' => $media_item->access('view label') ? $this->t('Remove @label', ['@label' => $media_item->label()]) : $this->t('Remove media'),
@@ -441,6 +491,19 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
           // Prevent errors in other widgets from preventing removal.
           '#limit_validation_errors' => $limit_validation_errors,
         ],
+        'edit_button' => [
+          '#type' => 'submit',
+          '#value' => $this->t('Edit'),
+          '#weight' => 1,
+          '#access' => $media_item->access('update'),
+          '#attributes' => [
+            'aria-label' => $media_item->access('view label') ? $this->t('Edit @label', ['@label' => $media_item->label()]) : $this->t('Edit media'),
+            'class' => ['edit-media', 'use-ajax'],
+            'href' => $media_item->toUrl('edit-form', ['query' => $state->all()])->toString(),
+            'data-dialog-type' => 'modal',
+            'data-dialog-options' => '{"width":"80%"}',
+          ],
+        ],
         'rendered_entity' => $preview,
         'target_id' => [
           '#type' => 'hidden',
@@ -462,52 +525,6 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
       ];
     }
 
-    $cardinality_unlimited = ($element['#cardinality'] === FieldStorageDefinitionInterface::CARDINALITY_UNLIMITED);
-    $remaining = $element['#cardinality'] - count($referenced_entities);
-
-    // Inform the user of how many items are remaining.
-    if (!$cardinality_unlimited) {
-      if ($remaining) {
-        $cardinality_message = $this->formatPlural($remaining, 'One media item remaining.', '@count media items remaining.');
-      }
-      else {
-        $cardinality_message = $this->t('The maximum number of media items have been selected.');
-      }
-
-      // Add a line break between the field message and the cardinality message.
-      if (!empty($element['#description'])) {
-        $element['#description'] .= '<br />';
-      }
-      $element['#description'] .= $cardinality_message;
-    }
-
-    // Create a new media library URL with the correct state parameters.
-    $selected_type_id = reset($allowed_media_type_ids);
-    $remaining = $cardinality_unlimited ? FieldStorageDefinitionInterface::CARDINALITY_UNLIMITED : $remaining;
-    // This particular media library opener needs some extra metadata for its
-    // \Drupal\media_library\MediaLibraryOpenerInterface::getSelectionResponse()
-    // to be able to target the element whose 'data-media-library-widget-value'
-    // attribute is the same as $field_widget_id. The entity ID, entity type ID,
-    // bundle, field name are used for access checking.
-    $entity = $items->getEntity();
-    $opener_parameters = [
-      'field_widget_id' => $field_widget_id,
-      'entity_type_id' => $entity->getEntityTypeId(),
-      'bundle' => $entity->bundle(),
-      'field_name' => $field_name,
-    ];
-    // Only add the entity ID when we actually have one. The entity ID needs to
-    // be a string to ensure that the media library state generates its
-    // tamper-proof hash in a consistent way.
-    if (!$entity->isNew()) {
-      $opener_parameters['entity_id'] = (string) $entity->id();
-
-      if ($entity->getEntityType()->isRevisionable()) {
-        $opener_parameters['revision_id'] = (string) $entity->getRevisionId();
-      }
-    }
-    $state = MediaLibraryState::create('media_library.opener.field_widget', $allowed_media_type_ids, $selected_type_id, $remaining, $opener_parameters);
-
     // Add a button that will load the Media library in a modal using AJAX.
     $element['open_button'] = [
       '#type' => 'button',
diff --git a/core/themes/claro/claro.theme b/core/themes/claro/claro.theme
index 6c24db719baac84d6d58f52e2fab979c6e478c6b..a7049d9fc78a0b85a3126c3a39e8bb048a957fef 100644
--- a/core/themes/claro/claro.theme
+++ b/core/themes/claro/claro.theme
@@ -1452,6 +1452,8 @@ function claro_preprocess_item_list__media_library_add_form_media_list(array &$v
  * This targets each media item selected in an entity reference field.
  */
 function claro_preprocess_media_library_item__widget(array &$variables): void {
+  $variables['content']['edit_button']['#attributes']['class'][] = 'media-library-item__edit';
+  $variables['content']['edit_button']['#attributes']['class'][] = 'icon-link';
   $variables['content']['remove_button']['#attributes']['class'][] = 'media-library-item__remove';
   $variables['content']['remove_button']['#attributes']['class'][] = 'icon-link';
 }
diff --git a/core/themes/claro/css/theme/media-library.css b/core/themes/claro/css/theme/media-library.css
index a8f12da1f0b5b75c8044591a05c45d8073b7d480..ffa8444278f5a367a8b1a09bc0ba68e6e9362cf7 100644
--- a/core/themes/claro/css/theme/media-library.css
+++ b/core/themes/claro/css/theme/media-library.css
@@ -750,6 +750,12 @@
 .media-library-item__edit,
 .media-library-item__edit:hover,
 .media-library-item__edit:focus,
+.media-library-item__edit.button,
+.media-library-item__edit.button:disabled,
+.media-library-item__edit.button:disabled:active,
+.media-library-item__edit.button:hover,
+.media-library-item__edit.button:focus,
+.media-library-item__edit.button:not(:focus),
 .media-library-item__remove,
 .media-library-item__remove:hover,
 .media-library-item__remove:focus,
@@ -792,7 +798,13 @@
   left: 0.625rem;
 }
 
-.media-library-item__edit {
+.media-library-item__edit,
+.media-library-item__edit.button,
+.media-library-item__edit.button:disabled,
+.media-library-item__edit.button:disabled:active,
+.media-library-item__edit.button:hover,
+.media-library-item__edit.button:focus,
+.media-library-item__edit.button:not(:focus) {
   /* !important to override button class border. */
   border: 1px solid var(--color-gray-200) !important;
   background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16'%3e%3cg fill='%23545560'%3e%3cpath d='M14.545 3.042l-1.586-1.585a1.003 1.003 0 00-1.414 0L10.252 2.75l3 3 1.293-1.293a1.004 1.004 0 000-1.415zM5.25 13.751l-3-3 6.998-6.998 3 3zM.908 14.775c-.087.262.055.397.316.312l2.001-.667-1.65-1.646-.667 2.001z'/%3e%3c/g%3e%3c/svg%3e");
@@ -801,7 +813,9 @@
   background-size: 0.75rem;
 }
 
-.media-library-item__edit:active {
+.media-library-item__edit:active,
+.media-library-item__edit.button:active,
+.media-library-item__edit.button:disabled:active {
   background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16'%3e%3cg%3e%3cpath fill='%23ffffff' d='M14.545 3.042l-1.586-1.585c-.389-.389-1.025-.389-1.414 0l-1.293 1.293 3 3 1.293-1.293c.389-.389.389-1.026 0-1.415z'/%3e%3crect fill='%23ffffff' x='5.129' y='3.8' transform='matrix(-.707 -.707 .707 -.707 6.189 20.064)' width='4.243' height='9.899'/%3e%3cpath fill='%23ffffff' d='M.908 14.775c-.087.262.055.397.316.312l2.001-.667-1.65-1.646-.667 2.001z'/%3e%3c/g%3e%3c/svg%3e");
 }
 
diff --git a/core/themes/claro/css/theme/media-library.pcss.css b/core/themes/claro/css/theme/media-library.pcss.css
index 0ac3e4b8bbf19fbc0cfb92c1ec75cc7bba481bdc..4c3b62977461126316adfb5201cc1978c7518759 100644
--- a/core/themes/claro/css/theme/media-library.pcss.css
+++ b/core/themes/claro/css/theme/media-library.pcss.css
@@ -696,6 +696,12 @@
 .media-library-item__edit,
 .media-library-item__edit:hover,
 .media-library-item__edit:focus,
+.media-library-item__edit.button,
+.media-library-item__edit.button:disabled,
+.media-library-item__edit.button:disabled:active,
+.media-library-item__edit.button:hover,
+.media-library-item__edit.button:focus,
+.media-library-item__edit.button:not(:focus),
 .media-library-item__remove,
 .media-library-item__remove:hover,
 .media-library-item__remove:focus,
@@ -736,7 +742,13 @@
   left: 10px;
 }
 
-.media-library-item__edit {
+.media-library-item__edit,
+.media-library-item__edit.button,
+.media-library-item__edit.button:disabled,
+.media-library-item__edit.button:disabled:active,
+.media-library-item__edit.button:hover,
+.media-library-item__edit.button:focus,
+.media-library-item__edit.button:not(:focus) {
   /* !important to override button class border. */
   border: 1px solid var(--color-gray-200) !important;
   background-image: url("../../images/icons/545560/pencil.svg");
@@ -744,7 +756,9 @@
   background-position: center;
   background-size: 12px;
 }
-.media-library-item__edit:active {
+.media-library-item__edit:active,
+.media-library-item__edit.button:active,
+.media-library-item__edit.button:disabled:active {
   background-image: url("../../../../misc/icons/ffffff/pencil.svg");
 }
 .media-library-item__remove,
