diff --git a/lib/Drupal/Core/Field/Plugin/Field/FieldType/StringItem.php b/lib/Drupal/Core/Field/Plugin/Field/FieldType/StringItem.php
index b51764c294..1c146298b3 100644
--- a/lib/Drupal/Core/Field/Plugin/Field/FieldType/StringItem.php
+++ b/lib/Drupal/Core/Field/Plugin/Field/FieldType/StringItem.php
@@ -72,8 +72,14 @@ class StringItem extends StringItemBase {
    */
   public static function generateSampleValue(FieldDefinitionInterface $field_definition) {
     $random = new Random();
-    $values['value'] = $random->word(mt_rand(1, $field_definition->getSetting('max_length')));
-    return $values;
+    $label = $field_definition->getLabel();
+    $value = "($label) ". $random->word(mt_rand(1, $field_definition->getSetting('max_length')));
+
+    $value = substr($value, 0, strlen($label) + 25);
+    $space_position = rand(strlen($label) + 9, strlen($label) + 15);
+    $value = substr($value, 0, $space_position) . ' ' . substr($value, $space_position);
+
+    return ['value' => $value];
   }
 
   /**
diff --git a/lib/Drupal/Core/Field/Plugin/Field/FieldType/StringLongItem.php b/lib/Drupal/Core/Field/Plugin/Field/FieldType/StringLongItem.php
index 93c6b27bcb..9c2ea5fd47 100644
--- a/lib/Drupal/Core/Field/Plugin/Field/FieldType/StringLongItem.php
+++ b/lib/Drupal/Core/Field/Plugin/Field/FieldType/StringLongItem.php
@@ -39,7 +39,7 @@ class StringLongItem extends StringItemBase {
    */
   public static function generateSampleValue(FieldDefinitionInterface $field_definition) {
     $random = new Random();
-    $values['value'] = $random->paragraphs();
+    $values['value'] = "({$field_definition->getLabel()}) " . $random->paragraphs(3);
     return $values;
   }
 
diff --git a/modules/text/src/Plugin/Field/FieldType/TextItemBase.php b/modules/text/src/Plugin/Field/FieldType/TextItemBase.php
index e70fadacb1..5e980b7ed0 100644
--- a/modules/text/src/Plugin/Field/FieldType/TextItemBase.php
+++ b/modules/text/src/Plugin/Field/FieldType/TextItemBase.php
@@ -76,11 +76,11 @@ abstract class TextItemBase extends FieldItemBase {
 
     if (empty($settings['max_length'])) {
       // Textarea handling
-      $value = $random->paragraphs();
+      $value = "({$field_definition->getLabel()}) " . $random->paragraphs(3);
     }
     else {
       // Textfield handling.
-      $value = substr($random->sentences(mt_rand(1, $settings['max_length'] / 3), FALSE), 0, $settings['max_length']);
+      $value = "({$field_definition->getLabel()}) " . substr($random->sentences(mt_rand(1, $settings['max_length'] / 3), FALSE), 0, $settings['max_length']);
     }
 
     $values = [
