diff --git a/src/Plugin/Field/FieldWidget/LocalistUrlWidget.php b/src/Plugin/Field/FieldWidget/LocalistUrlWidget.php
index f2d6526..93dbc63 100644
--- a/src/Plugin/Field/FieldWidget/LocalistUrlWidget.php
+++ b/src/Plugin/Field/FieldWidget/LocalistUrlWidget.php
@@ -344,13 +344,13 @@ class LocalistUrlWidget extends LinkWidget {
   /**
    * Get the data from the localist API.
    */
-  protected function getApiData() {
+  public function getApiData($base_url = NULL) {
     // Data was already fetched.
     if ($this->apiData) {
       return $this->apiData;
     }
 
-    $base_url = $this->getSetting('base_url');
+    $base_url = $base_url ?? $this->getSetting('base_url');
 
     // Check for some cached data before we fetch it all again.
     if ($cache = $this->cache->get("localist_api:$base_url")) {
@@ -364,13 +364,15 @@ class LocalistUrlWidget extends LinkWidget {
     }
 
     try {
-      $this->fetchApiData();
+      $this->fetchApiData($base_url);
     }
     catch (\Throwable $e) {
       if (!$this->apiData) {
         throw $e;
       }
     }
+
+    return $this->apiData;
   }
 
   /**
@@ -379,8 +381,7 @@ class LocalistUrlWidget extends LinkWidget {
    * @return array
    *   Keyed array of api data.
    */
-  protected function fetchApiData(): array {
-    $base_url = $this->getSetting('base_url');
+  protected function fetchApiData($base_url): array {
     $options = [
       'timeout' => 5,
       'base_uri' => $base_url,
@@ -402,7 +403,7 @@ class LocalistUrlWidget extends LinkWidget {
         continue;
       }
 
-      $this->apiData[$key] = $this->fetchPagedApiData($key, $response['page']['total']);
+      $this->apiData[$key] = $this->fetchPagedApiData($base_url, $key, $response['page']['total']);
     }
 
     $this->cache->set("localist_api:$base_url", [
@@ -424,8 +425,7 @@ class LocalistUrlWidget extends LinkWidget {
    * @return array
    *   Indexed array of api data.
    */
-  protected function fetchPagedApiData($endpoint, $total_count): array {
-    $base_url = $this->getSetting('base_url');
+  protected function fetchPagedApiData($base_url, $endpoint, $total_count): array {
     $options = [
       'timeout' => 5,
       'base_uri' => $base_url,
