diff --git a/modules/ui_patterns_library/src/Plugin/Deriver/LibraryDeriver.php b/modules/ui_patterns_library/src/Plugin/Deriver/LibraryDeriver.php
index a0da46e..fc9fb35 100644
--- a/modules/ui_patterns_library/src/Plugin/Deriver/LibraryDeriver.php
+++ b/modules/ui_patterns_library/src/Plugin/Deriver/LibraryDeriver.php
@@ -154,27 +154,14 @@ class LibraryDeriver extends AbstractYamlPatternsDeriver {
   /**
    * Create a list of all directories to scan.
    *
-   * This includes all module directories and directories of the default theme
-   * and all of its possible base themes.
+   * This includes all module and theme directories.
    *
    * @return array
    *   An array containing directory paths keyed by their extension name.
    */
-  protected function getDirectories() {
-    $default_theme = $this->themeHandler->getDefault();
-    $base_themes = $this->themeHandler->getBaseThemes($this->themeHandler->listInfo(), $default_theme);
-    $theme_directories = $this->themeHandler->getThemeDirectories();
-
-    $directories = [];
-    if (isset($theme_directories[$default_theme])) {
-      $directories[$default_theme] = $theme_directories[$default_theme];
-      foreach (array_keys($base_themes) as $name) {
-        $directories[$name] = $theme_directories[$name];
-      }
-    }
-
-    return $directories + $this->moduleHandler->getModuleDirectories();
-  }
+   protected function getDirectories() {
+     return $this->moduleHandler->getModuleDirectories() + $this->themeHandler->getThemeDirectories();
+   }
 
   /**
    * Get extension name that hosts the given YAML definition file.
