diff --git a/js/views_ajax_history.js b/js/views_ajax_history.js
index ca2eb3d..1c86b9c 100644
--- a/js/views_ajax_history.js
+++ b/js/views_ajax_history.js
@@ -111,7 +111,7 @@
         value = viewArgs[name];
       }
       if (Array.isArray(value)) {
-        $.merge(query, $.map($.unique(value), function (sub) {
+        $.merge(query, $.map($.uniqueSort(value), function (sub) {
           return name + '=' + encodeURIComponent(sub);
         }));
       }