diff --git a/src/Controller/WeMegaMenuAdminController.php b/src/Controller/WeMegaMenuAdminController.php
index 301f0ba..a6a8f05 100644
--- a/src/Controller/WeMegaMenuAdminController.php
+++ b/src/Controller/WeMegaMenuAdminController.php
@@ -2,6 +2,8 @@
 
 namespace Drupal\we_megamenu\Controller;
 
+use Drupal\Component\Utility\Xss;
+use Drupal\Core\Cache\Cache;
 use Drupal\Core\Url;
 use Drupal\Core\Controller\ControllerBase;
 use Drupal\we_megamenu\WeMegaMenuBuilder;
@@ -32,7 +34,7 @@ class WeMegaMenuAdminController extends ControllerBase {
    *
    * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
    *   The config factory.
-   * 
+   *
    * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
    *   The module handler.
    */
@@ -89,9 +91,9 @@ class WeMegaMenuAdminController extends ControllerBase {
     if (isset($_POST['action']) && $_POST['action'] == 'save') {
       $data_config = $_POST['data_config'];
       $theme = $_POST['theme'];
-      $menu_name = $_POST['menu_name'];
+      $menu_name = Xss::filter($_POST['menu_name']);
       WeMegaMenuBuilder::saveConfig($menu_name, $theme, $data_config);
-      we_megamenu_flush_render_cache();
+      Cache::invalidateTags(['config:system.menu.' . $menu_name]);
     }
     exit;
   }
@@ -104,7 +106,7 @@ class WeMegaMenuAdminController extends ControllerBase {
       $theme_array = WeMegaMenuBuilder::renderWeMegaMenuBlock($_POST['menu_name'], $_POST['theme']);
       $markup = render($theme_array);
       echo $markup;
-      we_megamenu_flush_render_cache();
+      Cache::invalidateTags(['config:system.menu.' . Xss::filter($_POST['menu_name'])]);
       exit;
     }
 
@@ -117,7 +119,7 @@ class WeMegaMenuAdminController extends ControllerBase {
       $theme_array = WeMegaMenuBuilder::renderWeMegaMenuBlock($_POST['menu_name'], $_POST['theme']);
       $markup = render($theme_array);
       echo $markup;
-      we_megamenu_flush_render_cache();
+      Cache::invalidateTags(['config:system.menu.' . Xss::filter($_POST['menu_name'])]);
       exit;
     }
     exit;
@@ -129,7 +131,7 @@ class WeMegaMenuAdminController extends ControllerBase {
   public function styleOfBackendWeMegaMenu() {
     if (isset($_POST['type'])) {
       \Drupal::state()->set('we_megamenu_backend_style', $_POST['type']);
-      we_megamenu_flush_render_cache();
+      Cache::invalidateTags(['config:system.menu.' . Xss::filter($_POST['menu_name'])]);
     }
     exit;
   }
diff --git a/src/WeMegaMenuBuilder.php b/src/WeMegaMenuBuilder.php
index eb5a2bc..715dff3 100644
--- a/src/WeMegaMenuBuilder.php
+++ b/src/WeMegaMenuBuilder.php
@@ -303,6 +303,7 @@ class WeMegaMenuBuilder {
       $query = Drupal::database()->select('we_megamenu', 'km');
       $query->addField('km', 'data_config');
       $query->condition('km.menu_name', $menu_name);
+      $query->condition('km.theme', $theme);
       $query->range(0, 1);
       $result = $query->execute()->fetchField();
       return json_decode($result);
@@ -440,7 +441,7 @@ class WeMegaMenuBuilder {
             }
           }
         }
-      } 
+      }
     }
   }
 
@@ -696,7 +697,7 @@ class WeMegaMenuBuilder {
   public static function repositionMenuItems($menu_name, $theme_name = '', $menu_config, $child_item) {
     $list_menu_items = WeMegaMenuBuilder::getMenuItems($menu_name);
 
-    
+
     \Drupal::logger('megamenu')->warning( print_r(json_encode($list_menu_items), true) );
     foreach ($list_menu_items as $uuid => $childs) {
     }
@@ -818,7 +819,7 @@ class WeMegaMenuBuilder {
     $new_item->rows_content = $rows_content;
     $new_item->submenu_config = $submenu_config;
     $new_item->item_config = $item_config;
-    return $new_item;   
+    return $new_item;
   }
 
   public static function createNewMegamenuCellItem($menu_item, $menu_id) {
@@ -868,7 +869,7 @@ class WeMegaMenuBuilder {
                 $menu_id = $menu_item['derivativeId'];
                 if($megamenu_mlid != $menu_id) {
                   $new_megamenu_item = self::createNewMegamenuCellItem($menu_item, $menu_id);
-                  
+
                   if(!isset($menu_config[$derivativeId]['rows_content'][$i][$j]['col_config'])) {
                     $new_col_config = self::createNewMegamenuColConfig();
                     $megamenu_config->menu_config->{$derivativeId}->rows_content[$i][$j]->col_config = $new_col_config;
@@ -882,7 +883,7 @@ class WeMegaMenuBuilder {
               }
             }
           }
-        }        
+        }
       }
     }
     else {
diff --git a/we_megamenu.module b/we_megamenu.module
index d95808c..2b151a9 100644
--- a/we_megamenu.module
+++ b/we_megamenu.module
@@ -168,7 +168,7 @@ function template_preprocess_we_megamenu_backend(array &$vars) {
       WeMegaMenuBuilder::updateMegamenuFromDrupalMenu($data_config, $menu_items, 0);
       $data_config->menu_update_flag = 0;
       WeMegaMenuBuilder::saveConfig($menu_name, $theme_name, json_encode($data_config));
-      // drupal_flush_all_caches();      
+      // drupal_flush_all_caches();
     }
   }
 
@@ -229,7 +229,7 @@ function template_preprocess_we_megamenu_frontend(&$vars) {
       WeMegaMenuBuilder::updateMegamenuFromDrupalMenu($data_config, $menu_items, 0);
       $data_config->menu_update_flag = 0;
       WeMegaMenuBuilder::saveConfig($menu_name, $theme_name, json_encode($data_config));
-      // drupal_flush_all_caches();      
+      // drupal_flush_all_caches();
     }
   }
 
@@ -809,11 +809,3 @@ function we_megamenu_preprocess_html(&$vars) {
     }
   }
 }
-
-/**
- * Function we_megamenu_flush_render_cache.
- */
-function we_megamenu_flush_render_cache() {
-  $renderCache = \Drupal::service('cache.render');
-  $renderCache->invalidateAll();
-}
